c45.py

```diff
@@ -0,0 +1,217 @@
+"""
+C4.5 Binary Decision Tree Implementation
+
+Usage: 
+Read csv file in; will be stored as a 2 Dimensional list. (See fread())
+Train a classifier (i.e. train(list))
+Prune the decision tree (i.e. prune_tree(tree, 0.5))
+Predict the result (i.e. predict([.....], classifier))
+
+The function assumes that the last column of your data is populated by labels.
+
+Example of usage:
+
+data = fread("./test_val_dump.csv", True)
+drop_first_col = []
+for x in data:
+    drop_first_col.append(x[1:])
+tree = train(drop_first_col)
+prune_tree(tree, 0.5)
+print(predict([2,0,2,6,1,0,0,3,1,0,0,2,0,0,0,0,0,1,.223606798,0,.285714,.141421,0,.253546],tree))
+
+"""
+
+from collections import OrderedDict
+from math import log
+import csv
+
+def n_distinct_dict(rows):
+    counts = dict()
+    for x in rows:
+        xs = x[-1]
+        if xs not in counts: 
+            #Add it to the counts dictionary
+            counts[xs] = 0
+        counts[xs] += 1
+    return counts
+
+def entropy(X):
+    """
+    Calculate Entropy (as per Octavian)
+    """
+    counts = n_distinct_dict(X)
+    log_2 = lambda x: log(x)/log(2)
+    #Declare entropy value
+    entropy = 0.0
+    
+    for c in counts:
+        #Calculate P(C_i)
+        p = float(counts[c])/len(X)
+        entropy = entropy -  p*log_2(p)
+    return entropy
+
+def gini(X):
+    """
+    Calculate Gini Index
+    """
+    total = len(X)
+    counts = n_distinct_dict(X)
+    imp = 0.0
+    
+    for k1 in counts:
+        p1 = float(counts[k1])/total  
+        for k2 in counts:
+            if k1 == k2: continue
+            p2 = float(counts[k2])/total
+            imp += p1*p2
+    return imp
+
+class Tree:
+    """
+    Decision Tree class
+    """
+    def __init__(self, col=-1, value=None, right_branch=None, left_branch=None, results=None):
+        self.col = col
+        self.value = value
+        self.right_branch = right_branch
+        self.left_branch = left_branch
+        self.results = results
+
+
```

Comment:
can you explain the least gain part; should there be a default value or should this be user-specified every time